src/ch09-02-recoverable-errors-with-result.md

```diff
@@ -391,146 +332,133 @@ don't want to include it for rustdoc testing purposes. -->
-function in one line. If we couldn’t use the `?` operator on `Option`, we’d
-have to implement this logic using more method calls or a `match` expression.
-
-Note that you can use the `?` operator on a `Result` in a function that returns
-`Result`, and you can use the `?` operator on an `Option` in a function that
-returns `Option`, but you can’t mix and match. The `?` operator won’t
-automatically convert a `Result` to an `Option` or vice versa; in those cases,
-you can use methods like the `ok` method on `Result` or the `ok_or` method on
-`Option` to do the conversion explicitly.
-
-So far, all the `main` functions we’ve used return `()`. The `main` function is
-special because it’s the entry and exit point of executable programs, and there
-are restrictions on what its return type can be for the programs to behave as
-expected.
-
-Luckily, `main` can also return a `Result<(), E>`. Listing 9-12 has the
-code from Listing 9-10 but we’ve changed the return type of `main` to be
-`Result<(), Box<dyn Error>>` and added a return value `Ok(())` to the end. This
-code will now compile:
+<span class="caption">Listing 9-11: Użycie operatora `?` na wartości `Option<T>`</span>
+
+Ta funkcja zwraca `Option<char>`, ponieważ możliwe jest, że jest tam znak, ale możliwe
+jest również, że go tam nie ma. Ten kod pobiera argument `text` z wycinka łańcucha
+i wywołuje na nim metodę `lines`, która zwraca iterator nad liniami w łańcuchu. 
+Ponieważ funkcja ta chce sprawdzić pierwszą linię, wywołuje `next` na iteratorze,
+aby uzyskać pierwszą wartość z iteratora. Jeśli `text` jest pustym łańcuchem,
+to wywołanie `next` zwróci `None`, w którym to przypadku użyjemy `?` by zatrzymać
+i zwrócić `None` z `last_char_of_first_line`. Jeśli `text` nie jest pustym łańcuchem,
+`next` zwróci wartość `Some` zawierającą wycinek łańcucha pierwszej linii w `text`.
+
+Znak `?` wyodrębnia wycinek łańcucha i możemy wywołać `chars` na tym wycinku łańcucha,
+aby uzyskać iterator jego znaków. Interesuje nas ostatni znak w tej pierwszej linii,
+więc wywołujemy `last`, aby zwrócić ostatni element w iteratorze. Jest to `Option`,
+ponieważ możliwe jest, że pierwsza linia jest pustym ciągiem, na przykład jeśli `text`
+zaczyna się od pustej linii, ale zawiera znaki w innych liniach, jak w `„\nhi”`. 
+Jednakże, jeśli w pierwszej linii znajduje się ostatni znak, zostanie on zwrócony
+w wariancie `Some`. Operator `?` w środku daje nam zwięzły sposób na wyrażenie tej
+logiki, pozwalając nam zaimplementować funkcję w jednej linii. Gdybyśmy nie mogli 
+użyć operatora `?` w `Option`, musielibyśmy zaimplementować tę logikę używając więcej
+wywołań metod lub wyrażenia `match`.
+
+Zauważ, że możesz użyć operatora `?` na `Result` w funkcji, która zwraca `Result` i możesz
+użyć operatora `?` na `Option` w funkcji, która zwraca `Option`, ale nie możesz mieszać
+i dopasowywać. Operator `?` nie konwertuje automatycznie `Result` na `Option` lub odwrotnie;
```

Comment:
"mieszać i dopasowywać" brzmi całkowicie niezrozumiale... może po prostu "ich mieszać"? Choć tak na  prawdę pewne mieszanie jest dozwolone, z tego co pamiętam to Result może zostać przekonwertowane do Option.
https://www.diki.pl/slownik-angielskiego?q=mix-and-match

Reply:
chyba lepiej będzie "nie można ich dowolnie mieszać"

Reply:
Zmieniłem.